sw/device/silicon_creator/rom/e2e/rom_e2e_c_init_test.c

```diff
@@ -0,0 +1,184 @@
+    CHECK(legal22 == 3);
+  }
+  // UART RX (input 42) -> IOC3 (MIO pad 25)
+  uint32_t insel_uartrx =
+      abs_mmio_read32(kPinmuxBase + PINMUX_MIO_PERIPH_INSEL_42_REG_OFFSET);
+  CHECK(insel_uartrx == kTopEarlgreyPinmuxInselIoc3);
+
+  // IOC4 (MIO pad 26) -> UART TX (output 42)
+  uint32_t outsel_ioc4 =
+      abs_mmio_read32(kPinmuxBase + PINMUX_MIO_OUTSEL_26_REG_OFFSET);
+  CHECK(outsel_ioc4 == kTopEarlgreyPinmuxOutselUart0Tx);
+}
+
+/**
+ * Check that the UART has been initialized correctly.
+ *
+ * Expected configuration:
+ * - `CTRL.TX` (TX enable) bit is true
+ * - `CTRL.RX` (RX enable) bit is false
+ * - `CTRL.SLPBK` (system loopback) bit is false
+ * - `CTRL.LLPBK` (line loopback) bit is false
+ * - `CTRL.NCO` is set to the precomputed `kUartNCOValue` for this device
+ * - `CTRL.PARITY_EN` (parity enable) bit is false
+ * - All interrupts are disabled
+ * - All FIFOs have been cleared (both Rx and Tx are empty)
+ */
+static void uart_init_test(void) {
+  // Check the control register values.
+  uint32_t ctrl = abs_mmio_read32(kUartBase + UART_CTRL_REG_OFFSET);
+  CHECK(bitfield_bit32_read(ctrl, UART_CTRL_TX_BIT) == true);
+  CHECK(bitfield_bit32_read(ctrl, UART_CTRL_RX_BIT) == false);
+  CHECK(bitfield_bit32_read(ctrl, UART_CTRL_SLPBK_BIT) == false);
+  CHECK(bitfield_bit32_read(ctrl, UART_CTRL_LLPBK_BIT) == false);
+  CHECK(bitfield_field32_read(ctrl, UART_CTRL_NCO_FIELD) == kUartNCOValue);
+  CHECK(bitfield_bit32_read(ctrl, UART_CTRL_PARITY_EN_BIT) == false);
+
+  // Check that all interrupts are disabled.
+  uint32_t intr_enable =
+      abs_mmio_read32(kUartBase + UART_INTR_ENABLE_REG_OFFSET);
+  CHECK(intr_enable == 0x0);
+
+  // Check that both FIFOs are idle and empty.
+  uint32_t status = abs_mmio_read32(kUartBase + UART_STATUS_REG_OFFSET);
+  CHECK(bitfield_bit32_read(status, UART_STATUS_TXIDLE_BIT) == true);
+  CHECK(bitfield_bit32_read(status, UART_STATUS_RXIDLE_BIT) == true);
+  CHECK(bitfield_bit32_read(status, UART_STATUS_TXEMPTY_BIT) == true);
+  CHECK(bitfield_bit32_read(status, UART_STATUS_RXEMPTY_BIT) == true);
+}
+
+/**
+ * Check that the CPUCTRL CSR has been initialized correctly.
+ *
+ * Expected configuration:
+ * - Bits 0:5 of the CPUCTRL CSR should match the value specified in OTP.
+ */
+static void cpuctrl_init_test(void) {
+  uint32_t expected_value =
+      otp_read32(OTP_CTRL_PARAM_CREATOR_SW_CFG_CPUCTRL_OFFSET);
+  bitfield_field32_t cpuctrl_mask = {.mask = 0x3f, .index = 0};
+  uint32_t cpuctrl_csr;
+  CSR_READ(CSR_REG_CPUCTRL, &cpuctrl_csr);
+  CHECK(bitfield_field32_read(cpuctrl_csr, cpuctrl_mask) == expected_value);
+}
+
+/**
+ * Test entrypoint.
+ *
+ * Because this function becomes the entrypoint for the manifest, it must be
+ * word-aligned, otherwise it will trigger a `kManifestBadEntryPoint` boot
+ * fault.
+ */
+void __attribute__((aligned(4))) _ottf_start(void) {
+  pinmux_init_test();
+  uart_init_test();
+  cpuctrl_init_test();
+}
```

Comment:
Consider the following so that it's clear that the passed.
```
LOG_INFO("PASS!");
wait_for_interrupt(); // //sw/lib/runtime:hart
```

Reply:
I seem to be getting a timeout when I switch to this method. Is there a reason it might not work when the test is not using OTTF? I don't seem to get printouts in general from `LOG_INFO`, only boot faults.

Reply:
Sorry for forgetting to mention this but since we're running without ottf_main, we need to initialize pinmux, uart, and the print library:

```
void __attribute__((aligned(4))) _ottf_start(void) {

    ...

    // Clobber peripherals to output a message
    pinmux_init();
    uart_init();
    base_set_stdout((buffer_sink_t){
      .data = NULL,
      .sink = uart_sink,
    });
    LOG_INFO("PASS!");
    wait_for_interrupt();
}
```